readthedocs/builds/tasks.py

```diff
@@ -453,3 +458,174 @@ def send_build_status(build_pk, commit, status, link_to_build=False):
+                settings.PRODUCTION_DOMAIN,
+                self.build.get_absolute_url(),
+            ),
+            'unsub_url': 'https://{}{}'.format(
+                settings.PRODUCTION_DOMAIN,
+                reverse('projects_notifications', args=[self.project.slug]),
+            ),
+        }
+
+        if self.build.commit:
+            title = _('Failed: {project[name]} ({commit})').format(
+                commit=self.build.commit[:8],
+                **context,
+            )
+        else:
+            title = _('Failed: {project[name]} ({version[verbose_name]})').format(
+                **context
+            )
+
+        log.info(
+            'Sending email notification. email=%s project=%s version=%s build=%s',
+            email, self.project.slug, self.version.slug, self.build.id,
+        )
+        send_email(
+            email,
+            title,
+            template='projects/email/build_failed.txt',
+            template_html='projects/email/build_failed.html',
+            context=context,
+        )
+
+    def send_webhook(self, webhook):
+        """
+        Send webhook notification.
+
+        The payload is signed using HMAC-SHA1,
+        for users to be able to verify the authenticity of the request.
+
+        Webhooks that don't have a payload,
+        are from the old implementation, for those we keep sending the
+        old default payload.
+
+        An HttpExchange object is created for each transaction.
+        """
+        payload = webhook.get_payload(
+            version=self.version,
+            build=self.build,
+            event=self.event,
+        )
+        if not payload:
+            # Default payload from old webhooks.
+            payload = json.dumps({
+                'name': self.project.name,
+                'slug': self.project.slug,
+                'build': {
+                    'id': self.build.id,
+                    'commit': self.build.commit,
+                    'state': self.build.state,
+                    'success': self.build.success,
+                    'date': self.build.date.strftime('%Y-%m-%d %H:%M:%S'),
+                },
+            })
+
+        headers = {
+            'content-type': 'application/json',
+            'User-Agent': f'Read-the-Docs/{__version__}'
+        }
+        if webhook.secret:
+            headers['X-Hub-Signature'] = webhook.sign_payload(payload)
+
+        try:
+            log.info(
+                'Sending webhook notification. webhook=%s project=%s version=%s build=%s',
+                webhook.pk, self.project.slug, self.version.slug, self.build.pk,
+            )
+            response = requests.post(
+                webhook.url,
+                data=payload,
+                headers=headers,
+                timeout=self.webhook_timeout,
+            )
+            HttpExchange.objects.from_requests_exchange(
+                response=response,
+                related_object=webhook,
+            )
```

Comment:
We could create this object in two steps, before sending the request and updating it after the request was made. This is so we save the exchange even if the request failed (like timeout).